File: EPI/14.10/swift/14.10.swift
public class Node {
    public let value: Int
    public var next: Node?

    public init(value: Int, next: Node?) {
        self.value = value
        self.next = next
    }

    public func sort() -> Node {
        if next == nil {
            return self
        }

        let mid = self.middle()
        let secondHalf = mid.next!
        mid.next = nil

        let sortedFirstHalf = self.sort()
        let sortedSecondHalf = secondHalf.sort()

        return sortedFirstHalf.merge(sortedSecondHalf)
    }

    func middle() -> Node {
        var slow = self
        var fast = self

        while let fnn = fast.next?.next {
            slow = slow.next!
            fast = fnn
        }

        return slow
    }

    func merge(_ other: Node) -> Node {
        let (first, second) = self.value < other.value ? (self, other) : (other, self)
        first.next = first.next?.merge(second) ?? second
        return first
    }
}

File: EPI/14.10/typescript/14.10.ts
export interface List<T> {
    value: T;
    next?: List<T>;
}

export function sort<T>(head: List<T>): List<T> {
    if (!head.next) {
        return head;
    }
    const mid = split(head);
    return merge(sort(head), sort(mid));
}

function split<T>(head: List<T>): List<T> {
    let beforeSlow = head;
    let slow = head.next!;
    let fast: List<T> | undefined = head.next;

    while (fast && fast.next) {
        beforeSlow = slow;
        slow = slow.next!;
        fast = fast.next.next;
    }

    beforeSlow.next = undefined;
    return slow;
}

// See 8.1 for an iterative version, as well as a more concise but not stable recursive version.
function merge<T>(a: List<T>, b: List<T>): List<T> {
    if (a.value <= b.value) {
        a.next = a.next ? merge(a.next, b) : b;
        return a;
    } else {
        b.next = b.next ? merge(a, b.next) : a;
        return b;
    }
}
